### PR TITLE
Support std::map in InputParameters

### LINKED ISSUE #14894: Let input parameters support std::map
## Reason
There are cases that having map style input can be useful. For example, `petsc_options_iname` and `petsc_options_value` can be merged into one parameter
```
params.addParam<std::map<std::string, std::string>>("petsc_options", {{"-pctype", "lu"}}, valid_keys, "Available PETSc options in input files");
```
Then in the input users can give
```
petsc_options = '-pctype lu'
```
as the command line arguments.
Valid keys is used to detect any input errors on typing in the keys. We should be able to detect errors like duplicated keys in the inputs. Of course, the parameter can be optional, i.e. the default parameter does not have to be provided.

This is just one example. In reactor analysis, we often need to give the initial atomic densities of a set of isotopes. Currently we have to ask users to have
```
initial_isotopes = 'H1 O16 U235'
initial_densities = '0.1 0.1 0.1'
```
It will be nice we can have something like
```
initial_isotopes =
'
  H1   0.1
  O16  0.1
  U235 0.1
'
```
This will make input cleaner especially when we have lots of isotopes.

## Design
Looks like supporting map in parameters is a non-trivial task. It may require changes in libMesh as well.

## Impact
An alternative on inputting parameters that can make inputs more user friendly.

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #14894

@YaqiWang can you give me a list of `std::pair<KeyType, MappedType>` that you will want? My first test is with `std::string - > Real`

### CODE REVIEW

File: framework/src/parser/Parser.C [hunk 1]
```diff
@@ -1236,6 +1236,14 @@ Parser::extractParams(const std::string & prefix, InputParameters & p)
           dynamic_cast<InputParameters::Parameter<std::vector<ptype>> *>(par),                     \
           in_global,                                                                               \
           global_params_block)
+#define setmap(key_type, mapped_type)                                                              \
```
Comment (rwcarlsen) [inline framework/src/parser/Parser.C:1239]: Just for good measure - maybe undef this with the other undefs below.

**Comment by YaqiWang** [comment]
Yes, typically we have `<std::string, Real>`. Also we want to support `<std::string, std::string>` and leave the conversion to users.

File: test/src/kernels/MapMultiplyCoupledVars.C [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  params.addRequiredCoupledVar("w", "The second coupled var");
+  params.addRequiredParam<std::map<std::string, Real>>("coupled_var_multipliers",
+                                                       "A map from variable names to multipliers");
+  return params;
+}
+
+MapMultiplyCoupledVars::MapMultiplyCoupledVars(const InputParameters & parameters)
+  : ADKernel(parameters), _v(adCoupledValue("v")), _w(adCoupledValue("w"))
+{
+  const auto & coupled_map = getParam<std::map<std::string, Real>>("coupled_var_multipliers");
```
Comment (YaqiWang) [inline test/src/kernels/MapMultiplyCoupledVars.C:28]: can we move this line into above header? I want to have a show-case that we can construct a const reference member variable like other parameters.

File: framework/src/parser/Parser.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/parser/Parser.C:1239]: :+1:

File: test/src/kernels/MapMultiplyCoupledVars.C (hunk 2 above)
Comment (lindsayad) [reply test/src/kernels/MapMultiplyCoupledVars.C]: k done

**Comment by YaqiWang** [comment]
Just to make sure if `<std::string, std::string>` is supported or not?

**Comment by lindsayad** [comment]
> Just to make sure if `<std::string, std::string>` is supported or not?

Yea. So far we have:
```
      setmap(string, Real);
      setmap(string, string);
```
And we can always add more

**Comment by lindsayad** [comment]
This is ready for review

**Review by rwcarlsen — APPROVED** [review]

**Comment by lindsayad** [comment]
@YaqiWang you good with us merging?

**Review by YaqiWang — APPROVED** [review]

**Pull request merged**